### ISSUE TITLE
BNS namespace price function cannot be changed after NAMESPACE_REVEAL

### ISSUE THREAD

**Issue by MarvinJanssen** [blockstack/stacks-blockchain#2092]
**Is your feature request related to a problem? Please describe.**
Feature request per discussion with @lgalabru. One can only set the price function of a namespace only once; i.e., during the `NAMESPACE_REVEAL` phase. It could make domain names prohibitively expensive in the long run. Namespaces that are created shortly after the Stacks 2.0 launch will arguably be most susceptible to this issue. Say we want to price a domain name to cost roughly US$5, and this translates into a specific pricing function where f(_name_) results in _n_ STX ≈ US$5, it goes without saying that it is unlikely for the STX value to stay the same over time.

**Describe the solution you'd like**
A way to update the price function, perhaps via a `NAMESPACE_UPDATE` transaction. However, such an option can be dangerous as it could allow the namespace owner to arbitrarily raise and lower the prices. I assume the design choice to fix the price function hinges on this fact. Therefore, it is probably more interesting to have the option to let a smart contract decide the price of a domain name by implementing a trait. A namespace owner can then create a price function that evolves over time (for example, a function of the block height), creating a voting mechanism, or even deploy a price oracle. It opens the door to innovation. The trait implementation itself will still be transparent to the user, who can then make the informed decision on whether to purchase a domain name in a particular namespace.

**Describe alternatives you've considered**
An overhaul of BNS, a story for another time. 

**Additional context**
https://github.com/blockstack/stacks-blockchain/blob/7a2d79b7b03d03615decb26eb4b639b1cf8ba345/src/chainstate/stacks/boot/bns.clar#L351

**Comment by lgalabru** [comment]
Thank you for opening this issue @MarvinJanssen!

@jcnelson @kantai what's our take on this feature request?

On the overhaul of BNS mentioned above, I think that @MarvinJanssen had something like this: https://github.com/blockstack/stacks-blockchain/issues/1801 in mind.

**Comment by MarvinJanssen** [comment]
It is indeed quite similar to #1801, thank you for pointing it out to me. Each domain can be handled by a smart contract, which then handles the registration of sub domains under its name, which in turn can be more smart contracts, thus creating a chain. The "namespaces" no longer receive special treatment and instead become top level domains like ordinary DNS. The BNS root contract simply contains a mapping of all top level domains and their corresponding contracts, and functionality to register a new TLD at the root. Name functions (resolution, registration, etc.) are passed to the proper contract. All contracts should naturally implement the proper traits. If domains are made NFTs, they can even be traded.

I think you could have one chainable Clarity implementation for all of it.

**Comment by jcnelson** [comment]
We can make it so that a namespace, when created, can have an option to allow the creator to update the price function later.  A user would be able to tell if a particular namespace had this property or not, and choose a namespace accordingly in which to register names.

**Comment by MarvinJanssen** [comment]
That seems like a good intermediary solution to me.

**Comment by diwakergupta** [comment]
@MarvinJanssen if you're able to help out with this, we'd welcome a PR. Otherwise we'll try to get to it after December 15th, but it will be lower priority.

**Comment by MarvinJanssen** [comment]
I'll get on it.

**Comment by MarvinJanssen** [comment]
After some discussions with Gina and others we concluded that having a `NAMESPACE_UPDATE` feature as mentioned is probably not the right way to go. We feel it would put too much power in the hands of the namespace owner to change the price function at will. Instead, more advanced price functions are best reserved for BNSv3. (Like the aforementioned overhaul.) Thoughts?

**Comment by diwakergupta** [comment]
> After some discussions with Gina and others we concluded that having a `NAMESPACE_UPDATE` feature as mentioned is probably not the right way to go. We feel it would put too much power in the hands of the namespace owner to change the price function at will. Instead, more advanced price functions are best reserved for BNSv3. (Like the aforementioned overhaul.) Thoughts?

@MarvinJanssen thanks, in that case we're no longer planning this for Stacks 2.0. There is a path to introduce this in a new BNS contract later if needed. Alternatively, we can always pull this in for Stacks 2.1.

Feel free to close this out unless there's some other actionable task here.

**Comment by MarvinJanssen** [comment]
Sounds good 👍.
